### ISSUE TITLE
Having trouble understanding a passage

### ISSUE THREAD

**Issue by sgjohnson1981** [MicrosoftDocs/cpp-docs#2600]
Can someone please explain the following passage in plain English? Is this "optimizing" necessary? Does removing it keep compatibility for say Linux? I checked the "more information" link. Didn't really help explain things for me.

"The new project template for a DLL project adds PROJECTNAME_EXPORTS to the defined preprocessor macros. In this example, Visual Studio defines MATHLIBRARY_EXPORTS when your MathLibrary DLL project is built.

When the MATHLIBRARY_EXPORTS macro is defined, the MATHLIBRARY_API macro sets the __declspec(dllexport) modifier on the function declarations. This modifier tells the compiler and linker to export a function or variable from the DLL for use by other applications. When MATHLIBRARY_EXPORTS is undefined, for example, when the header file is included by a client application, MATHLIBRARY_API applies the __declspec(dllimport) modifier to the declarations. This modifier optimizes the import of the function or variable in an application. For more information, see dllexport, dllimport."

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 08c574a2-cef0-ab13-7375-55851f966ef6
* Version Independent ID: fe3421c7-e9f4-4986-589f-fadaedce583e
* Content: [Walkthrough: Create and use your own Dynamic Link Library (C++)](https://docs.microsoft.com/en-us/cpp/build/walkthrough-creating-and-using-a-dynamic-link-library-cpp?view=msvc-160)
* Content Source: [docs/build/walkthrough-creating-and-using-a-dynamic-link-library-cpp.md](https://github.com/Microsoft/cpp-docs/blob/master/docs/build/walkthrough-creating-and-using-a-dynamic-link-library-cpp.md)
* Product: **visual-cpp**
* Technology: **cpp-tools**
* GitHub Login: @corob-msft
* Microsoft Alias: **corob**

**Comment by corob-msft** [comment]
@sgjohnson1981

Shared code libraries on Linux don't work quite the same way as DLLs on Windows, and the `__declspec(dllimport)` and `__declspec(dllexport)` attributes don't apply outside of the Windows DLL context. These keywords are specific to Microsoft platforms. You'll want to use preprocessor macros to make them disappear on other platforms in your cross-platform code. Describing how to do that is way outside the scope of this article.

The `dllimport` keyword tells the compiler to make an optimization that saves one level of indirection when a client accesses an exported DLL data item or function. Without `dllimport`, client code would have to make an indirect call to the exported function through a table of export addresses that's created at load time, when the DLL is loaded in the client's process space. With `dllimport`, the compiler creates code to fixup the client calls to point directly at the exports. Your client code will work fine either way, but it's slightly faster when you use `dllimport`. This explanation is also outside the scope of this particular article, so I'm not convinced there's something to update here.

**Comment by corob-msft** [comment]
There's also a [useful](https://devblogs.microsoft.com/oldnewthing/20060720-20/?p=30453) [series](https://devblogs.microsoft.com/oldnewthing/20060721-06/?p=30433) of [articles](https://devblogs.microsoft.com/oldnewthing/20060724-00/?p=30403) on [The Old New Thing](https://devblogs.microsoft.com/oldnewthing/) blog that might explain this better. The linker plumbing for DLLs is complicated, and not easy to explain in simple terms.
